File: DFS&BFS/python/1525.py
from collections import deque
import sys
input = sys.stdin.readline
arr = [list(map(int,input().split())) for _ in range(3)]
answer = -1
def make_string (arr):
  s = ""
  for i in range(3):
    s += "".join(map(str,arr[i]))
  return s

def range_check (r,c):
  if r<0 or c<0 or r>=3 or c>=3:
    return False
  return True

def zero_search(arr):
  for i in range(3):
    for j in range(3):
      if arr[i][j] == 0:
        return [i,j]

moves = [(-1,0),(0,1),(1,0),(0,-1)]
q = deque()
s = make_string(arr)
if s != "123456780":
  q.append((arr,0))
else:
  answer = 0

visited = set()
while q:
  arr_,cnt = q.popleft()
  r,c = zero_search(arr_)
  flag = False
  for move in moves:
    nr,nc = r + move[0], c+move[1]
    if not range_check(nr,nc): continue
    next_arr = [[arr_[i][j] for j in range(3)] for i in range(3)]
    next_arr[r][c], next_arr[nr][nc] = next_arr[nr][nc], next_arr[r][c]
    s = make_string(next_arr)
    if s == "123456780":
      answer = cnt+1
      flag = True
      break
    if s in visited:
      continue
    visited.add(s)
    q.append((next_arr,cnt+1))
  if flag:
    break
print(answer)

File: DFS&BFS/javascript/1525.js
// 노드js는 메모리초과, 파이썬 풀이는 정답판정

const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

arr = [];
for (let i = 0; i < 3; i++) {
  arr.push(input[i].split(" ").map(Number));
}

let flag = false;
const makeString = (arr) => {
  str = "";
  for (let i = 0; i < 3; i++) {
    str += arr[i].join("");
  }
  return str;
};
const moves = [
  [-1, 0],
  [0, 1],
  [1, 0],
  [0, -1],
];

const q = [];
let idx = 0;
let answer = -1;
const visited = new Set();
let s = makeString(arr);
visited.add(s);
if (s !== "123456780") {
  q.push([arr, 0]);
}

const rangeCheck = (r, c) => {
  if (r < 0 || c < 0 || r >= 3 || c >= 3) return false;
  return true;
};
const check = (arr, cnt, r, c) => {
  for (let move of moves) {
    let [nr, nc] = [r + move[0], c + move[1]];
    if (rangeCheck(nr, nc)) {
      const newArr = [[], [], []];
      for (let i = 0; i < 3; i++) {
        for (let j = 0; j < 3; j++) {
          newArr[i].push(arr[i][j]);
        }
      }
      [newArr[r][c], newArr[nr][nc]] = [newArr[nr][nc], newArr[r][c]];
      const s = makeString(newArr);
      if (s === "123456780") {
        flag = true;
        answer = cnt + 1;
        return true;
      }
      if (visited.has(s)) {
        continue;
      }
      visited.add(s);
      q.push([newArr, cnt + 1]);
    }
  }
};

while (q.length > idx) {
  [newArr, cnt] = q[idx++];
  for (let i = 0; i < 3; i++) {
    for (let j = 0; j < 3; j++) {
      if (newArr[i][j] === 0) {
        check(newArr, cnt, i, j);
      }
    }
    if (flag) {
      break;
    }
  }
  if (flag) {
    break;
  }
}
console.log(answer);
